README.md
# ヴァナ語辞典 #
 iPhone用「FF11用語辞典 ～ ウィンダスの仲間たち版」ブラウザ

## ヴァナ語辞典とは ##
「FF11用語辞典 ～ ウィンダスの仲間たち版」を iPhone で快適に見るためのブラウザです。  
Titanium Mobile を使って作られていますので、Android でも動くかもしれません。

### 主な機能 ###
* 用語検索
* ブックマーク
* 閲覧履歴

## 謝辞 ##
「FF11用語辞典 ～ ウィンダスの仲間たち版」の管理者様、記事を更新されているユーザの皆様に感謝いたします。

## ライセンス ##
Copyright © 2010 Takatoshi -morimori- MORIYAMA.  
MIT-LICENSE.ja をお読みください。


ChangeLog.txt
2010-12-27 Takatoshi -morimori- MORIYAMA
* version 0.9


Resources/app.js
// this sets the background color of the master UIView (when there are no windows/tab groups on it)
Ti.UI.setBackgroundColor('#000');

Ti.include('data_manager.js');

// create tab group
var tabGroup = Ti.UI.createTabGroup();
var searchWin = Ti.UI.createWindow({
  url: 'search.js',
  title:'検索',
  backgroundColor:'#fff',
  orientationModes: [Ti.UI.PORTRAIT, Ti.UI.UPSIDE_PORTRAIT, Ti.UI.LANDSCAPE_LEFT, Ti.UI.LANDSCAPE_RIGHT]
});
var searchTab = Ti.UI.createTab({
  icon: Ti.UI.iPhone.SystemIcon.SEARCH,
  title:'検索',
  window:searchWin
});
tabGroup.addTab(searchTab);

var bookmarkWin = Ti.UI.createWindow({
  url: 'bookmark.js',
  title:'ブックマーク',
  backgroundColor:'#fff',
  orientationModes: [Ti.UI.PORTRAIT, Ti.UI.UPSIDE_PORTRAIT, Ti.UI.LANDSCAPE_LEFT, Ti.UI.LANDSCAPE_RIGHT]
});
var bookmarkTab = Ti.UI.createTab({
  icon: Ti.UI.iPhone.SystemIcon.BOOKMARKS,
  title:'ブックマーク',
  window:bookmarkWin
});

tabGroup.addTab(bookmarkTab);

var historyWin = Ti.UI.createWindow({
  url: 'history.js',
  title:'履歴',
  backgroundColor:'#fff',
  orientationModes: [Ti.UI.PORTRAIT, Ti.UI.UPSIDE_PORTRAIT, Ti.UI.LANDSCAPE_LEFT, Ti.UI.LANDSCAPE_RIGHT]
});
var historyTab = Ti.UI.createTab({
  icon: Ti.UI.iPhone.SystemIcon.HISTORY,
  title:'履歴',
  window:historyWin
});

tabGroup.addTab(historyTab);

var infoWin = Ti.UI.createWindow({
  url: 'info.js',
  title:'このアプリについて',
  backgroundColor:'#fff',
  orientationModes: [Ti.UI.PORTRAIT, Ti.UI.UPSIDE_PORTRAIT, Ti.UI.LANDSCAPE_LEFT, Ti.UI.LANDSCAPE_RIGHT]
});
var infoTab = Ti.UI.createTab({
  title:'About',
  window:infoWin
});

tabGroup.addTab(infoTab);

tabGroup.open();


Resources/search.js
var context = 'search';

Ti.include('utils/jquery-wrapper.js', 'browser.js', 'data_manager.js');

var win = Ti.UI.currentWindow;

var searchForm = Ti.UI.createSearchBar({
  showCancel: true,
  height: 44,
  top: 0
});
win.add(searchForm);

var tableView = Ti.UI.createTableView({
  rowHeight: 45,
  top: 44
});
tableView.addEventListener('click', function(event) {
  var browser = new Browser(event.rowData.wordId);
});
win.add(tableView);

function addResultRow(id, title, read) {
  var rowCanvas = Ti.UI.createView({
    width: 300,
    height: 44,
    backgroundColor: '#ffffff',
    opaque: 0
  });

  var titleLabel = Ti.UI.createLabel({
    left: 20,
    top: 3,
    width: 260,
    height: 21,
    text: title,
    font: {fontFamily: 'Helvetica', fontSize: 16},
    color: '#000000'
  });
  rowCanvas.add(titleLabel);

  var readLabel = Ti.UI.createLabel({
    left: 20,
    top: 22,
    width: 260,
    height: 17,
    text: read,
    font: {fontFamily: 'Helvetica', fontSize: 12},
    color: '#888888'
  });
  rowCanvas.add(readLabel);

  var row = Ti.UI.createTableViewRow({
    width: 320,
    height: 44,
    backgroundColor: '#ffffff',
    hasChild: true,
    wordId: id
  });
  row.add(rowCanvas);

  tableView.appendRow(row);
}

searchForm.addEventListener('return', function(event) {
  var indicator = Ti.UI.createActivityIndicator();
  win.setRightNavButton(indicator);
  indicator.show();
  var proxy = Ti.UI.createWebView();
  var callback = function(event) {
    proxy.removeEventListener('load', callback);
    indicator.hide();
    win.setRightNavButton(null);
    var editList = proxy.evalJS('$("#content .editlist").html();');
    if (editList != '') {
      var id = editList.match(/wiki\.cgi\?.*?id=(\d+)/)[1];
      var read = proxy.evalJS('$("#content .title > .yomi").text();').match(/^\((.+)\)$/)[1];
      proxy.evalJS('$("#content .title > .yomi").remove();');
      var title = proxy.evalJS('$("#content .title").text();');
      addResultRow(id, title, read);
      browser.open(id);
    } else {
      var result = proxy.evalJS('$("#content ul").html();');
      if (result == '') { return; }
      $.each(result.replace(/<li>/g, '').replace(/\n/g, '').split('</li>'), function(index, value) {
        var data = value.match(/^<a href="http:\/\/wiki\.ffo\.jp\/html\/(\d+)\.html">([^<]+)<\/a> \/ (.+)$/);
        if (data) { addResultRow(data[1], data[2], data[3]); }
      });
    }
  };
  tableView.data = [];
  proxy.addEventListener('load', callback);
  proxy.url = 'http://wiki.ffo.jp/search.cgi?CCC=愛&Command=Search&qf=' + searchForm.value + '&order=match&ffotype=title&type=title';
  proxy.reload();
  searchForm.blur();
});

searchForm.addEventListener('cancel', function(event) {
  searchForm.blur();
});


Resources/browser.js
var template = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'container.html').read().text;

var Browser = function(id) {
  this.id = id;
  this.window = Ti.UI.createWindow({
    // title: title,
    backgroundColor: 'white',
    orientationModes: [Ti.UI.PORTRAIT, Ti.UI.UPSIDE_PORTRAIT, Ti.UI.LANDSCAPE_LEFT, Ti.UI.LANDSCAPE_RIGHT]
  });
  this.window.addEventListener('close', $.proxy(function(event) {
    Ti.API.log(this.window.title + ' CLOSED');
  }, this));
  this.window.addEventListener('focus', $.proxy(function(event) {
    Ti.API.log(this.window.title + ' FOCUSED');
  }, this));
  this.window.addEventListener('blur', $.proxy(function(event) {
    Ti.API.log(this.window.title + ' BLURED');
  }, this));
  var indicator = Ti.UI.createActivityIndicator();
  this.window.setRightNavButton(indicator);
  indicator.show();
  Ti.UI.currentTab.open(this.window, {animated: true});

  this.proxy = Ti.UI.createWebView({url: 'http://wiki.ffo.jp/html/' + id + '.html'});
  this.proxy.addEventListener('load', $.proxy(this._proxyOnLoad, this));
  this.proxy.reload();
};

Browser.prototype = {
  _proxyOnLoad: function(event) {
    var yomi = this.proxy.evalJS('$("#content .title > .yomi").text();');
    var match = yomi.match(/^\((.+)\)$/);
    if (match == null) { Ti.API.log(this.proxy.html); this.proxy.reload(); }
    var read = this.proxy.evalJS('$("#content .title > .yomi").text();').match(/^\((.+)\)$/)[1];
    this.proxy.evalJS('$("#content .title > .yomi").remove();');
    this.window.title = this.proxy.evalJS('$("#content .title").text();');
    this.proxy.evalJS('$("#content .content a.edit").remove();');
    this.proxy.evalJS('$.each($("#content .content a[href]"), function(i,v){ $(v).attr({"onClick": "return clickHook(this, \'' + context + '\');"}); });');
    var browserView = Ti.UI.createWebView({html: template.replace(/<body><\/body>/, '<body>' + this.proxy.evalJS('$("#content .content").html();') + '</body>')});
    this.window.add(browserView);
    var indicator = this.window.getRightNavButton();
    indicator.hide();
    var bookmarkBtn = Ti.UI.createButton({enabled: !dm.isBookmarked(this.id), wordId: this.id, systemButton: Ti.UI.iPhone.SystemButton.BOOKMARKS});
    var win = this.window;
    bookmarkBtn.addEventListener('click', $.proxy(function(event) {
      if (this.enabled) {
        dm.addBookmark(this.wordId);

        var label = Ti.UI.createLabel({text: 'ブックマークしました', color: '#ffffff', font: {fontFamily: 'Helvetica', fontSize: 14, fontWeight: 'bold'}});
        win.setToolbar([label], {animated: true});
        setTimeout(function() {
          win.setToolbar(null,{animated: true});
        }, 2500);
      }
      this.enabled = !this.enabled;
    }, bookmarkBtn));
    this.window.setRightNavButton(bookmarkBtn);
    dm.updateHistory(this.id, this.window.title, read);
    this.proxy = null;
  } 
};

Ti.App.addEventListener(context + 'LinkClick', function(event) {
  var data = event.url.match(/\/?(\d+)\.html$/);
  var browser = new Browser(data[1]);
});


Resources/history.js
var context = 'history';

Ti.include('utils/jquery-wrapper.js', 'browser.js', 'data_manager.js');

var win = Ti.UI.currentWindow;

var tableView = Ti.UI.createTableView({rowHeight: 45, editable: true});
tableView.addEventListener('click', function(event) {
  var browser = new Browser(event.rowData.wordId);
});
tableView.addEventListener('delete', function(event) {
  dm.removeHistory(event.rowData.wordId);
});
win.add(tableView);

var clearBtn = Ti.UI.createButton({systemButton: Titanium.UI.iPhone.SystemButton.TRASH});
clearBtn.addEventListener('click', function(event) {
  var alertDialog = Titanium.UI.createAlertDialog({
    title: '履歴削除',
    message: '全て削除してよろしいですか？',
    buttonNames: ['はい', 'いいえ']
  });
  alertDialog.addEventListener('click', function(event) {
    if (event.index == 0) {
      dm.clearHistories();
      tableView.data = [];

      var label = Ti.UI.createLabel({text: '履歴を削除しました', color: '#ffffff', font: {fontFamily: 'Helvetica', fontSize: 14, fontWeight: 'bold'}});
      win.setToolbar([label], {animated: true});
      setTimeout(function() {
        win.setToolbar(null,{animated: true});
      }, 2500);
    }
  });
  alertDialog.show();
});
win.setRightNavButton(clearBtn);

win.addEventListener('focus', function(event) {
  tableView.data = [];
  $.each(dm.histories('last_viewed_at DESC'), function(i,v) {
    var rowCanvas = Ti.UI.createView({
      width: 300,
      height: 44,
      backgroundColor: '#ffffff',
      opaque: 0
    });
  
    var titleLabel = Ti.UI.createLabel({
      left: 20,
      top: 3,
      width: 260,
      height: 21,
      text: v.title,
      font: {fontFamily: 'Helvetica', fontSize: 16},
      color: '#000000'
    });
    rowCanvas.add(titleLabel);
  
    var readLabel = Ti.UI.createLabel({
      left: 20,
      top: 22,
      width: 260,
      height: 17,
      text: v.read,
      font: {fontFamily: 'Helvetica', fontSize: 12},
      color: '#888888'
    });
    rowCanvas.add(readLabel);
  
    var row = Ti.UI.createTableViewRow({
      width: 320,
      height: 44,
      backgroundColor: '#ffffff',
      hasChild: true,
      wordId: v.id
    });
    row.add(rowCanvas);
  
    tableView.appendRow(row);
  });
});


Resources/bookmark.js
var context = 'bookmark';

Ti.include('utils/jquery-wrapper.js', 'browser.js', 'data_manager.js');

var win = Ti.UI.currentWindow;

var tableView = Ti.UI.createTableView({rowHeight: 45, editable: true});
tableView.addEventListener('click', function(event) {
  var browser = new Browser(event.rowData.wordId);
});
tableView.addEventListener('delete', function(event) {
  dm.removeBookmark(event.rowData.wordId);
});
win.add(tableView);

win.addEventListener('focus', function(event) {
  tableView.data = [];
  $.each(dm.bookmarks(), function(i,v) {
    var rowCanvas = Ti.UI.createView({
      width: 300,
      height: 44,
      backgroundColor: '#ffffff',
      opaque: 0
    });
  
    var titleLabel = Ti.UI.createLabel({
      left: 20,
      top: 3,
      width: 260,
      height: 21,
      text: v.title,
      font: {fontFamily: 'Helvetica', fontSize: 16},
      color: '#000000'
    });
    rowCanvas.add(titleLabel);
  
    var readLabel = Ti.UI.createLabel({
      left: 20,
      top: 22,
      width: 260,
      height: 17,
      text: v.read,
      font: {fontFamily: 'Helvetica', fontSize: 12},
      color: '#888888'
    });
    rowCanvas.add(readLabel);
  
    var row = Ti.UI.createTableViewRow({
      width: 320,
      height: 44,
      backgroundColor: '#ffffff',
      hasChild: true,
      wordId: v.id
    });
    row.add(rowCanvas);
  
    tableView.appendRow(row);
  });
});


Resources/data_manager.js
var DataManager = function() {
  this.db = Ti.Database.open('data.db');
  this.db.execute('CREATE TABLE IF NOT EXISTS histories (id INTEGER PRIMARY KEY, title VARCHAR(255) NOT NULL, read VARCHAR(255) NOT NULL, last_viewed_at INTEGER NOT NULL, view_count INTEGER NOT NULL);');
  this.db.execute('CREATE TABLE IF NOT EXISTS bookmarks (id INTEGER PRIMARY KEY, title VARCHAR(255) NOT NULL, read VARCHAR(255) NOT NULL, last_viewed_at INTEGER NOT NULL, view_count INTEGER NOT NULL);');
};

DataManager.prototype = {
  histories: function(order) {
    var rs = this.db.execute('SELECT id, title, read, last_viewed_at, view_count FROM histories ORDER BY ' + order);
    var result = [];
    if (rs.rowCount > 0) {
      var record;
      do {
        result.push({
          id:              rs.fieldByName('id'),
          title:           rs.fieldByName('title'),
          read:            rs.fieldByName('read'),
          last_visited_at: rs.fieldByName('last_viewed_at'),
          view_count:      rs.fieldByName('view_count')
        });
      } while (rs.next());
    }
    rs.close();
    return result;
  },
  updateHistory: function(id, title, read) {
    var rs = this.db.execute('SELECT id FROM histories WHERE id = ? LIMIT 1', id);
    if (rs.rowCount == 0) {
      rs.close();
      this.db.execute("INSERT INTO histories VALUES (?, ?, ?, strftime('%s','now'), ?)", id, title, read, 1);
    } else {
      rs.close();
      this.db.execute("UPDATE histories SET last_viewed_at = strftime('%s','now'), view_count = view_count + 1 WHERE id = ?", id);
    }
  },
  removeHistory: function(id) {
    this.db.execute('DELETE FROM histories WHERE id = ?', id);
  },
  clearHistories: function() {
    this.db.execute('DELETE FROM histories');
  },

  bookmarks: function() {
    var rs = this.db.execute('SELECT id, title, read FROM bookmarks ORDER BY title');
    var result = [];
    if (rs.rowCount > 0) {
      var record;
      do {
        result.push({
          id:    rs.fieldByName('id'),
          title: rs.fieldByName('title'),
          read:  rs.fieldByName('read')
        });
      } while (rs.next());
    }
    rs.close();
    return result;
  },
  addBookmark: function(id) {
    this.db.execute('INSERT INTO bookmarks SELECT * FROM histories WHERE id = ?', id);
  },
  removeBookmark: function(id) {
    this.db.execute('DELETE FROM bookmarks WHERE id = ?', id);
  },
  isBookmarked: function(id) {
    var rs = this.db.execute('SELECT id FROM bookmarks WHERE id = ? LIMIT 1', id);
    var count = rs.rowCount;
    rs.close();
    return count == 1;
  }
};

var dm = new DataManager();


Resources/info.js
var win = Ti.UI.currentWindow;

var titleLabel = Ti.UI.createLabel({
  text: 'ヴァナ語辞典',
  textAlign: 'center',
  top: 5,
  width: 'auto',
  height: 'auto',
  font: {fontFamily: 'Helvetica', fontSize: 16, fontWeight: 'bold'},
  color: '#000000'
});

var summaryLabel = Ti.UI.createLabel({
  text: 'FF11用語辞典ブラウザ',
  textAlign: 'center',
  top: 30,
  width: 'auto',
  height: 'auto',
  font: {fontFamily: 'Helvetica', fontSize: 12},
  color: '#444444'
});

var versionLabel = Ti.UI.createLabel({
  text: 'Version ' + Ti.App.version,
  textAlign: 'center',
  top: 45,
  width: 'auto',
  height: 25,
  font: {fontFamily: 'Helvetica', fontSize: 12},
  color: '#444444'
});

var aboutRow = Ti.UI.createTableViewRow({height: 'auto', touchEnabled: false});
aboutRow.add(titleLabel);
aboutRow.add(summaryLabel);
aboutRow.add(versionLabel);

var aboutSection = Ti.UI.createTableViewSection({touchEnabled: false});
aboutSection.add(aboutRow);


var poweredBySection = Ti.UI.createTableViewSection({headerTitle: 'Powered by', touchEnabled: false});
poweredBySection.add(Ti.UI.createTableViewRow({height: 'auto', title: 'FF11用語辞典 ～ ウィンダスの仲間たち版', fontSize: 12}));
poweredBySection.add(Ti.UI.createTableViewRow({height: 'auto', title: 'Titanium Mobile', fontSize: 12}));
poweredBySection.add(Ti.UI.createTableViewRow({height: 'auto', title: 'jQuery', fontSize: 12}));

var tableView = Ti.UI.createTableView({
  data: [aboutSection, poweredBySection],
  style: Titanium.UI.iPhone.TableViewStyle.GROUPED
});

win.add(tableView);
